### ISSUE TITLE
git checkout doesn't fail cleanly

### ISSUE THREAD

**Issue by jeffeb3** [github/git-lfs#1195]
TL;DR: When git lfs failes to fetch a file, the entire git command fails, and it doesn't leave the repo in the same state as it started in.

    > git clone https://github.com/jeffeb3/lfs_test.git
    Cloning into 'lfs_test'...
    remote: Counting objects: 29, done.
    remote: Compressing objects: 100% (23/23), done.
    remote: Total 29 (delta 6), reused 24 (delta 4), pack-reused 0
    Unpacking objects: 100% (29/29), done.
    Checking connectivity... done.
    Downloading Image.jpg (4.93 KB)
    Downloading blue.png (1.38 KB)
    Downloading green.png (24.46 KB)

    > git status
    On branch master
    Your branch is up-to-date with 'origin/master'.
    
    nothing to commit, working directory clean

Here, I disconnect my connection to github.com (by pulling out the Ethernet cable)

    > git checkout HEAD^
    Downloading green.png (1.38 KB)
    Error downloading object: green.png           (d6991edfa7ab08200f68151eda1283ffa2839c721c401304acb744c9502b4c63)
    
    Errors logged to /home/jeffe/tmp/test/lfs_test/.git/lfs/objects/logs/20160503T104313.4193124.log
    Use `git lfs logs last` to view the log.
    error: external filter git-lfs smudge -- %f failed 2
    error: external filter git-lfs smudge -- %f failed
    fatal: green.png: smudge filter lfs failed

    > git status
    On branch master
    Your branch is up-to-date with 'origin/master'.
    
    Changes not staged for commit:
      (use "git add/rm <file>..." to update what will be committed)
      (use "git checkout -- <file>..." to discard changes in working directory)
    
            deleted:    green.png

   no changes added to commit (use "git add" and/or "git commit -a")
 
So, first off. Obviously lfs can't get the file. You're not magic.

Secondly though, if the operation fails, then lfs needs to leave it in the beginning state. This is really the bug here.

Thirdly, and this maybe belongs in it's own feature, I'd rather at least have the option of letting checkouts like this succeed, leaving just the pointer around instead of the file that failed to fetch. I definitely do work when I can't get to the server, and I'd like to be able to still do all the funky git stuff (branch, checkout by hash, cherry-pick, rebase, etc) when not connected to the server. I can make lfs.fetchexclude="*" and then do all that work, as a workaround.

    git lfs version
    git-lfs/1.2.0 (GitHub; linux amd64; go 1.5.3)

P.S. I'm assuming the log isn't useful for you, because we know why it failed. Also the steps to reproduce it are above. I'll save it anyway though.

**Comment by technoweenie** [comment]
The smudge command is a strange case. We could spit out errors and then return the LFS pointer content to Git. Your Git repository won't be complete, but it's probably be better than messing up your git index.

It seems like part of the problem here is relying on git's smudge feature. If you disable that and run `git lfs pull` afterwards manually, then it should work better for you. The downloads will happen in parallel, so they'll be faster. Any errors won't disrupt the git index, leading to unwanted changed. This should work _today_.

**Comment by jeffeb3** [comment]
I don't know if most users would want the entire thing to end up right where I started (clean master), or end up as close to the end goal as possible, with pointer files in place of files where there were failures. I would certainly prefer the latter. Maybe that's the easier way to fix it.

I don't know anything about smudge (except for what I've read about it after reading your comment). The lfs smudge must be returning some kind of error to git to revert back to master. But for some reason the original file is deleted, and it's not getting put back. Maybe it's a bug in git. It seems like git should be aborting, going back to master (the file should be the master pointer) and smudging each file from the master's checkout, including the one that lfs knows how to find, which should put it back to a clean master state.

**Comment by sinbad** [comment]
In the smudge filter, if a download fails we do actually still write the pointer to the stdout. However this never makes it to `green.png` (since it's marked as deleted) which makes me think that when git sees a non-zero return from the smudge filter, that content is not flushed to the destination file. If it was then probably everything would be well - although the exit code does halt the operation mid-flight which would prevent you performing other git operations like merging. I agree that it would be really nice to allow people to do these operations offline if they wanted, so long as they were happy with the lack of content.

The reason we had to start returning non-zero from the smudge filter is that it's not possible to detect failure in automated scripts, CI etc without it. What would be really nice is for the exit code not to happen until the end of the process, but git doesn't give us that option. As Rick said it's possible to disable the smudge filters to bypass this; however having the smudge filter in place lets us automatically handle a lot of corner cases like diffing, reverting etc, which 'just work' with filters in place but go wrong if they're turned turned off (without manual handling after the fact). 

I've talked about this before, but I still think it would help to have a suite of convenience functions akin to `git lfs clone` but for the other main cases (pull, checkout branch, merge) which both speed up the process and gather errors into one place. But, that doesn't solve the problem when people use core git (which they're going to).

I'm wondering whether we can do something with the `post-checkout` hook - maybe making the smudge filter return 0 but also queue up an error for that hook to pick up. I've no idea if that will work in this case or whether we can rely on it in all cases (we'd have to be able to since I don't think git gives the smudge filter a context to know what top-level git call is being made).

**Comment by sinbad** [comment]
Just to follow up, I've been doing some testing with other hooks and there's too many gaps to avoid using a smudge filter.

`post-checkout` can mop-up after switching branches and general checkout operations, and if you return a non-zero result from that it does bubble-up to `git checkout`. But, none of `git reset`, `git pull`, `git merge` or `git rebase` call it. So those leave gaps in the working copy.

You can use `post-merge` and `post-rebase` to close some of those gaps, but `git reset` remains as a hole which is a major one. Also even when you use `post-merge` etc, which does get called on `git pull`, returning a non-zero error *does not* propagate to the enclosing `git` command so you have no way of signalling failure. 

I think we're basically stuck with the smudge filter if we want a system which behaves consistently, and maybe adding more wrapper commands for efficiency / nicety. I'll see if I can figure out / work around the smudge fail case leaving the file contents missing in the working copy as a coping mechanism if possible.

**Comment by sinbad** [comment]
Interesting point of note: the smudge problem will only occur if the file in question was modified between your current HEAD and the commit you're checking out. If it was added or removed (whether directly if going forward in history, or the opposite effect when going backwards e.g. going back to a commit before the file existed) there is no problem. 

When the smudge filter fails, the checkout is aborted and you're still on the same commit as when you started. Modified files are missing because git pre-deletes the existing copy of the file before calling the smudge filter, so that it can stream the new contents to it. On failure though, that stream output is aborted, leaving the file missing. Here's the git code that does it from `convert.c`:

```c
	if (finish_async(&async)) {   // <--- This fails when we return non-zero from smudge filter
		error("external filter %s failed", cmd);
		ret = 0;
	}

	if (ret) { // <--- This doesn't happen so whatever we write to the stream is ignored
		strbuf_swap(dst, &nbuf);
	}
```

It's interesting to note that even *if* git didn't discard what we write to the stream in this case, and the pointer content got written to `green.png`, in the specific case here the working copy would still show as dirty, because `green.png` would be marked as modified. Even though the clean filter deals with pointer content correctly, so the file being the placeholder rather than real content would not show as dirty, when the checkout fails the working copy actually remains on the same commit (in this case `master` and not `HEAD^`). If our pointer content had got to the working copy, it would be the pointer as at `HEAD^` because that's what git was telling us to checkout, which is not correct for `master` so would show modified. 

I think this will take a change to git to fix. We can't even take a wilder approach to fixing it, like issuing our own `git checkout HEAD` from the smudge filter in the fail case to put the unchanged version back, because the smudge filter can be called in a lot of contexts and we can't know what 'undo' action is appropriate in the fail case. Really rather than just binning out, maybe git should try to do some recovery action? I guess the trouble is in our case it's simple when the connection is just not available, but in a general case it may have binned in the middle or with other unknown context. Hmm.

**Comment by jeffeb3** [comment]
I agree with you (at least I agree with what I think you're saying). This seems like an issue in git.

Git deletes the file, sends the HEAD^ version to the smudge filter, which returns with an error code.

Git aborts, goes back to master. Leaves the file deleted...

It seems like they probably won't want to cache what they delete (although that would work), but they should check out the master version of the file, and smudge that (which in this case, git-lfs-smudge should succeed, because it's in my local storage).

Some wild speculation on my part can explain why they might not want to do that, but it's probably more likely that no one has complained. I'm not sure how many people are using smudge, but most of them probably recognize this problem when it happens and know how to fix it. That's some mediocre speculation on my part.

To get the "feature" part of what I was asking, is there some way to configure the smudge filter to not return an error code? Maybe something where the script folks can use --batch to get it to return non-zero? As I mentioned, the easiest workaround I've seen is to just fetchexclude everything while I'm doing that kind of offline branch management, and then I can un-fetchexclude when I'm done.

**Comment by sinbad** [comment]
I can totally see why git would just want to abort and let the user figure it out, there are too many variations that could potentially happen, from everyone's perspective. 

Having a "don't abort on fail" option in the smudge filter is a fair workaround until someone thinks of a better approach, it would need to be an environment variable or a gitconfig setting since smudge can be invoked in all kinds of contexts. Maybe both is best for full flexibility, that's what git tends to do.
